File: First year/Second Semester/Graphs/A1/A1 - C++/src/ui.cpp
#include "ui.h"

UI::UI(): graph(0), originalGraph(0){
}

void UI::run() {
    cout << "\"~~~ Graph Algorithms in C++! ~~~ \n";

    map<string, function<void()>> inputOptions = {
            {"1", [this] { this->GetNumberOfVertices(); }},
            {"2", [this] { this->PrintTheGraph(); }},
            {"3", [this] { this->PrintTheVerticesOfTheGraph(); }},
            {"4", [this] { this->CheckIfEdgeExistsBetweenTwoVertices(); }},
            {"5", [this] { this->GetInDegreeAndOutDegreeOfAVertex(); }},
            {"6", [this] { this->PrintTheInboundEdgesOfAVertex(); }},
            {"7", [this] { this->PrintTheOutboundEdgesOfAVertex(); }},
            {"8", [this] { this->GetCostOfAnEdge(); }},
            {"9", [this] { this->ModifyCostOfAnEdge(); }},
            {"10", [this] { this->AddVertex(); }},
            {"11", [this] { this->RemoveVertex(); }},
            {"12", [this] { this->AddEdge(); }},
            {"13", [this] { this->RemoveEdge(); }},
            {"copy", [this] { this->CreateACopyOfTheCurrentGraph(); }},
            {"restore", [this] { this->RestoreTheGraphToTheCopy(); }},
            {"create", [this] { this->CreateARandomGraph(); }},
            {"read", [this] { this->ReadFromFile(); }},
            {"write", [this] { this->WriteToFile(); }},
            {"exit", [] { UI::ExitApp(); }},
    };

    while (true) {
        PrintMenu();
        cout << "\nEnter a command: ";
        string command;
        cin >> command;
        try {
            inputOptions.at(command)();
        } catch (const out_of_range&) {
            cout << "Invalid command\n";
        }
    }
}

void UI::PrintMenu() {
    cout << "1. Get the number of vertices" << endl;
    cout << "2. Print the graph" << endl;
    cout << "3. Print the vertices of the graph" << endl;
    cout << "4. Given two vertices, find out whether there is an edge from the first one to the second one" << endl;
    cout << "5. Get the in degree and out degree of a vertex" << endl;
    cout << "6. Print the inbound edges of a vertex" << endl;
    cout << "7. Print the outbound edges of a vertex" << endl;
    cout << "8. Get the cost of an edge" << endl;
    cout << "9. Modify the cost of an edge" << endl;
    cout << "10. Add a vertex" << endl;
    cout << "11. Remove a vertex" << endl;
    cout << "12. Add an edge" << endl;
    cout << "13. Remove an edge" << endl;
    cout << "copy. Create a copy of the current graph" << endl;
    cout << "restore. Restore the graph to the copy" << endl;
    cout << "create. Create a random graph" << endl;
    cout << "read. Read a graph from a file" << endl;
    cout << "write. Write the graph to a file" << endl;
    cout << "exit. Exit the application" << endl;
}

void UI::GetNumberOfVertices() {
    cout << "The number of vertices is: " << graph.GetNumberOfVertices() << endl;
}

void UI::PrintTheGraph() {
    cout << "The graph is: \n" << graph.toString() << endl;
}

void UI::PrintTheVerticesOfTheGraph() {
    cout << "The vertices of the graph are: ";
    for (int vertex : graph.GetVertices()) {
        cout << vertex << " ";
    }
    cout << endl;
}

void UI::CheckIfEdgeExistsBetweenTwoVertices() {
    int source, destination;
    cout << "Enter the source vertex: ";
    cin >> source;
    cout << "Enter the destination vertex: ";
    cin >> destination;

    try {
        if (graph.CheckIfEdgeExists(source, destination)) {
            cout << "There is an edge from " << source << " to " << destination << endl;
        } else {
            cout << "There is no edge from " << source << " to " << destination << endl;
        }
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::GetInDegreeAndOutDegreeOfAVertex() {
    int vertex;
    cout << "Enter the vertex: ";
    cin >> vertex;

    try {
        int in_degree = graph.GetInboundEdges(vertex).size();
        int out_degree = graph.GetOutboundEdges(vertex).size();
        cout << "The in degree of " << vertex << " is " << in_degree << " and the out degree of " << vertex << " is " << out_degree << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::PrintTheInboundEdgesOfAVertex() {
    int vertex;
    cout << "Enter the vertex: ";
    cin >> vertex;

    try {
        vector<int> inboundEdges = graph.GetInboundEdges(vertex);
        cout << "The inbound edges of " << vertex << " are: ";
        for (int edge : inboundEdges) {
            cout << edge << " ";
        }
        cout << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::PrintTheOutboundEdgesOfAVertex() {
    int vertex;
    cout << "Enter the vertex: ";
    cin >> vertex;

    try {
        vector<int> outboundEdges = graph.GetOutboundEdges(vertex);
        cout << "The outbound edges of " << vertex << " are: ";
        for (int edge : outboundEdges) {
            cout << edge << " ";
        }
        cout << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::GetCostOfAnEdge() {
    int source, destination;
    cout << "Enter the source vertex: ";
    cin >> source;
    cout << "Enter the destination vertex: ";
    cin >> destination;

    try {
        int cost = graph.GetCostOfAnEdge(source, destination);
        cout << "The cost of the edge from " << source << " to " << destination << " is " << cost << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::ModifyCostOfAnEdge() {
    int source, destination, cost;
    cout << "Enter the source vertex: ";
    cin >> source;
    cout << "Enter the destination vertex: ";
    cin >> destination;
    cout << "Enter the new cost: ";
    cin >> cost;

    try {
        graph.SetCostOfAnEdge(source, destination, cost);
        cout << "The cost of the edge from " << source << " to " << destination << " has been modified to " << cost << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::AddVertex() {
    int vertex;
    cout << "Enter the vertex: ";
    cin >> vertex;

    try {
        graph.AddVertex(vertex);
        cout << "The vertex " << vertex << " has been added to the graph" << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::RemoveVertex() {
    int vertex;
    cout << "Enter the vertex: ";
    cin >> vertex;

    try {
        graph.RemoveVertex(vertex);
        cout << "The vertex " << vertex << " has been removed from the graph" << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::AddEdge() {
    int source, destination, cost;
    cout << "Enter the source vertex: ";
    cin >> source;
    cout << "Enter the destination vertex: ";
    cin >> destination;
    cout << "Enter the cost of the edge: ";
    cin >> cost;

    try {
        graph.AddEdge(source, destination, cost);
        cout << "The edge from " << source << " to " << destination << " with cost " << cost << " has been added to the graph" << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::RemoveEdge() {
    int source, destination;
    cout << "Enter the source vertex: ";
    cin >> source;
    cout << "Enter the destination vertex: ";
    cin >> destination;

    try {
        graph.RemoveEdge(source, destination);
        cout << "The edge from " << source << " to " << destination << " has been removed from the graph" << endl;
    } catch (const GraphException& e) {
        cout << e.what() << endl;
    }
}

void UI::CreateACopyOfTheCurrentGraph() {
    if (!originalGraph.CheckIfGraphIsEmpty()) {
        cout << "There is already a copy of the graph" << endl;
        return;
    }

    originalGraph = graph;
    shared_ptr<DirectedGraph> copied_graph = graph.GetCopyOfGraph();
    graph = *copied_graph;
    cout << "The current graph has been copied" << endl;
}

void UI::RestoreTheGraphToTheCopy() {
    if (originalGraph.CheckIfGraphIsEmpty()) {
        cout << "There is no copy of the graph" << endl;
        return;
    }

    graph = originalGraph;
    originalGraph.SetNumberOfVertices(0);
    cout << "The graph has been restored to the copy" << endl;
}

void UI::CreateARandomGraph() {
    int number_of_vertices, number_of_edges;
    cout << "Enter the number of vertices: ";
    cin >> number_of_vertices;
    cout << "Enter the number of edges: ";
    cin >> number_of_edges;

    try {
        graph.CreateRandomGraph(number_of_vertices, number_of_edges);
        cout << "A random graph with " << number_of_vertices << " vertices and " << number_of_edges << " edges has been created" << endl;
    } catch (const GraphException& ge) {
        cout << ge.what() << endl;
    }
}

void UI::ReadFromFile() {
    string file_name, file_path;
    cout << "Enter the file name: ";
    cin >> file_name;

    file_path = "../data/" + file_name;

    try {
        graph.ReadGraphFromFile(file_path);
        cout << "The graph has been read from the file " << file_name << endl;
    } catch (const GraphException& ge) {
        cout << ge.what() << endl;
    }
}

void UI::WriteToFile() {
    string file_name, file_path;
    cout << "Enter the file name: ";
    cin >> file_name;

    file_path = "../data/" + file_name;

    graph.WriteGraphToFile(file_path);
    cout << "The graph has been written to the file " << file_name << endl;
}

void UI::ExitApp() {
    cout << "Goodbye!" << endl;
    exit(EXIT_SUCCESS);
}

File: First year/Second Semester/Graphs/A1/A1 - python/ui.py
from directed_graph import DirectedGraph
from graph_exceptions import GraphException
from colorama import Fore


class UI:
    def __init__(self):
        self.__graph = DirectedGraph()
        self.__original_graph = None

    def run_app(self):
        print(Fore.MAGENTA + "~~~ 🌟🐍 Graph Algorithms in Python! 🐍🌟 ~~~\n")

        input_options = {
            "1": self.__get_number_of_vertices,
            "2": self.__print_the_graph,
            "3": self.__print_the_vertices_of_the_graph,
            "4": self.__check_if_edge_exists_between_two_vertices,
            "5": self.__get_in_degree_and_out_degree_of_a_vertex,
            "6": self.__print_the_inbound_edges_of_a_vertex,
            "7": self.__print_the_outbound_edges_of_a_vertex,
            "8": self.__get_cost_of_an_edge,
            "9": self.__modify_cost_of_an_edge,
            "10": self.__add_vertex,
            "11": self.__remove_vertex,
            "12": self.__add_edge,
            "13": self.__remove_edge,
            "copy": self.__create_a_copy_of_the_current_graph,
            "restore": self.__restore_the_graph_to_the_copy,
            "create": self.__create_a_random_graph,
            "read": self.__read_from_file,
            "write": self.__write_to_file,
            "exit": UI.__exit_app
        }

        while True:
            UI.print_menu()
            command = input(Fore.GREEN + "Enter a command: ")
            try:
                input_options[command]()
            except KeyError:
                print("Invalid command")

    @staticmethod
    def print_menu():
        print(Fore.LIGHTYELLOW_EX + "1.", Fore.BLUE + "Get the number of vertices")
        print(Fore.LIGHTYELLOW_EX + "2.", Fore.BLUE + "Print the graph")
        print(Fore.LIGHTYELLOW_EX + "3.", Fore.BLUE + "Print the vertices of the graph")
        print(Fore.LIGHTYELLOW_EX + "4.", Fore.BLUE + "Given two vertices, find out whether there is an edge from the first one to the second one")
        print(Fore.LIGHTYELLOW_EX + "5.", Fore.BLUE + "Get the in degree and out degree of a vertex")
        print(Fore.LIGHTYELLOW_EX + "6.", Fore.BLUE + "Print the inbound edges of a vertex")
        print(Fore.LIGHTYELLOW_EX + "7.", Fore.BLUE + "Print the outbound edges of a vertex")
        print(Fore.LIGHTYELLOW_EX + "8.", Fore.BLUE + "Get the cost of an edge")
        print(Fore.LIGHTYELLOW_EX + "9.", Fore.BLUE + "Modify the cost of an edge")
        print(Fore.LIGHTYELLOW_EX + "10.", Fore.BLUE + "Add a vertex")
        print(Fore.LIGHTYELLOW_EX + "11.", Fore.BLUE + "Remove a vertex")
        print(Fore.LIGHTYELLOW_EX + "12.", Fore.BLUE + "Add an edge")
        print(Fore.LIGHTYELLOW_EX + "13.", Fore.BLUE + "Remove an edge")
        print(Fore.LIGHTYELLOW_EX + "copy.", Fore.BLUE + "Create a copy of the current graph")
        print(Fore.LIGHTYELLOW_EX + "restore.", Fore.BLUE + "Restore the graph to the copy")
        print(Fore.LIGHTYELLOW_EX + "create.", Fore.BLUE + "Create a random graph")
        print(Fore.LIGHTYELLOW_EX + "read.", Fore.BLUE + "Read a graph from a file")
        print(Fore.LIGHTYELLOW_EX + "write.", Fore.BLUE + "Write the graph to a file")
        print(Fore.LIGHTYELLOW_EX + "exit.", Fore.BLUE + "Exit the application")

    def __get_number_of_vertices(self):
        print(f" The number of vertices is: {self.__graph.get_number_of_vertices()}")

    def __print_the_graph(self):
        print("The graph is: ")
        print(self.__graph)

    def __print_the_vertices_of_the_graph(self):
        print(f"The vertices of the graph are: {self.__graph.get_vertices()}")

    def __check_if_edge_exists_between_two_vertices(self):
        source = int(input("Enter the source vertex: "))
        destination = int(input("Enter the destination vertex: "))
        try:
            if self.__graph.check_if_edge_exists(source, destination):
                print(f"There is an edge from {source} to {destination}")
            else:
                print(f"There is no edge from {source} to {destination}")
        except GraphException:
            print(f"There is no edge from {source} to {destination}")

    def __get_in_degree_and_out_degree_of_a_vertex(self):
        vertex = int(input("Enter the vertex: "))
        try:
            in_degree = self.__graph.get_inbound_edges_degree(vertex)
            out_degree = self.__graph.get_outbound_edges_degree(vertex)
            print(f"The in degree of {vertex} is {in_degree} and the out degree of {vertex} is {out_degree}")
        except KeyError:
            print(f"The vertex {vertex} is not in the current graph.")

    def __print_the_inbound_edges_of_a_vertex(self):
        vertex = int(input("Enter the vertex: "))
        try:
            print(f"The inbound edges of {vertex} are: {self.__graph.inbound_edges[vertex]}")
        except KeyError:
            print(f"The vertex {vertex} is not in the current graph.")

    def __print_the_outbound_edges_of_a_vertex(self):
        vertex = int(input("Enter the vertex: "))
        try:
            print(f"The outbound edges of {vertex} are: {self.__graph.outbound_edges[vertex]}")
        except KeyError:
            print(f"The vertex {vertex} is not in the current graph.")

    def __get_cost_of_an_edge(self):
        source = int(input("Enter the source vertex: "))
        destination = int(input("Enter the destination vertex: "))
        try:
            print(f"The cost of the edge from {source} to {destination} is {self.__graph.get_cost_of_an_edge(source, destination)}")
        except GraphException:
            print(f"There is no edge from {source} to {destination}")

    def __modify_cost_of_an_edge(self):
        source = int(input("Enter the source vertex: "))
        destination = int(input("Enter the destination vertex: "))
        cost = int(input("Enter the new cost: "))
        try:
            self.__graph.set_cost_of_an_edge(source, destination, cost)
            print(f"The cost of the edge from {source} to {destination} has been modified to {cost}")
        except GraphException:
            print(f"There is no edge from {source} to {destination}")

    def __add_vertex(self):
        vertex = int(input("Enter the vertex: "))
        try:
            self.__graph.add_vertex(vertex)
            print(f"The vertex {vertex} has been added to the graph")
        except GraphException:
            print(f"The vertex {vertex} is already in the graph")

    def __remove_vertex(self):
        vertex = int(input("Enter the vertex: "))
        try:
            self.__graph.remove_vertex(vertex)
            print(f"The vertex {vertex} has been removed from the graph")
        except GraphException:
            print(f"The vertex {vertex} is not in the graph")

    def __add_edge(self):
        source = int(input("Enter the source vertex: "))
        destination = int(input("Enter the destination vertex: "))
        cost = int(input("Enter the cost of the edge: "))
        try:
            self.__graph.add_edge(source, destination, cost)
            print(f"The edge from {source} to {destination} with cost {cost} has been added to the graph")
        except GraphException:
            print(f"The edge from {source} to {destination} with cost {cost} is already in the graph")

    def __remove_edge(self):
        source = int(input("Enter the source vertex: "))
        destination = int(input("Enter the destination vertex: "))
        try:
            self.__graph.remove_edge(source, destination)
            print(f"The edge from {source} to {destination} has been removed from the graph")
        except GraphException:
            print(f"The edge from {source} to {destination} is not in the graph")

    def __create_a_copy_of_the_current_graph(self):
        if self.__original_graph is not None:
            print("There is already a copy of the graph")
            return

        self.__original_graph = self.__graph
        self.__graph = self.__graph.get_copy_of_graph()
        print("The current graph has been copied")

    def __restore_the_graph_to_the_copy(self):
        if self.__original_graph is None:
            print("There is no copy of the graph")
            return

        self.__graph = self.__original_graph
        self.__original_graph = None
        print("The graph has been restored to the copy")

    def __create_a_random_graph(self):
        number_of_vertices = int(input("Enter the number of vertices: "))
        number_of_edges = int(input("Enter the number of edges: "))
        try:
            self.__graph.create_random_graph(number_of_vertices, number_of_edges)
        except GraphException as ge:
            print(ge)
            return
        print(f"A random graph with {number_of_vertices} vertices and {number_of_edges} edges has been created")

    def __read_from_file(self):
        file_name = input("Enter the file name: ")
        try:
            self.__graph.read_graph_from_file(file_name)
        except GraphException as ge:
            print(ge)
            return
        print(f"The graph has been read from the file {file_name}")

    def __write_to_file(self):
        file_name = input("Enter the file name: ")
        self.__graph.write_graph_to_file(file_name)
        print(f"The graph has been written to the file {file_name}")

    @staticmethod
    def __exit_app():
        print(Fore.MAGENTA + "Goodbye! 🌟🐍")
        exit()
